Error handling improvements

coreymckrill opened the pull request:
* Rename 0-slack.php to 0-error-handling.php, since that file's only
  purpose is to load the Slack library for error handling purposes
  anyway.
* Move all of the error handling stuff from wcorg-misc.php to
  0-error-handling.php. There's enough of it to warrant a separate file,
  and this way it all gets loaded before the rest of the mu-plugins
  so it can potentially catch errors there as well.
* Run the custom error handler and shutdown function in other
  environments besides production, as long as the SANDBOX_SLACK_USERNAME
  constant is defined. This way the errors will be sent to the sandbox
  owner's Slack account, rather than the log channel.
* In the shutdown function, only send the error to Slack if it is one
  of the types of errors that can't be handled in the custom error
  handler.

coreymckrill committed:
Error handling improvements

* Rename 0-slack.php to 0-error-handling.php, since that file's only
  purpose is to load the Slack library for error handling purposes
  anyway.
* Move all of the error handling stuff from wcorg-misc.php to
  0-error-handling.php. There's enough of it to warrant a separate file,
  and this way it all gets loaded before the rest of the mu-plugins
  so it can potentially catch errors there as well.
* Run the custom error handler and shutdown function in other
  environments besides production, as long as the SANDBOX_SLACK_USERNAME
  constant is defined. This way the errors will be sent to the sandbox
  owner's Slack account, rather than the log channel.
* In the shutdown function, only send the error to Slack if it is one
  of the types of errors that can't be handled in the custom error
  handler.

iandunn commented:
I like the idea of moving this all the that file, but I think we might want it to load earlier than mu-plugins, like during `define-wordcamp.php`, which is the equivalent of where w.org loads theirs. That way it'll catch things that happen earlier in the process.

iandunn commented:
Maybe we can store the file somewhere in `wp-content` though, so it can be in the public repo, and just include it from `define`?

vedanshujain reviewed (approved):
LGTM. Other than Ian's comment.

coreymckrill commented:
I'm not sure that the `add_action` and `wp_schedule_event` functions would be available when *define-wordcamp* is loading. I'll do some testing.

iandunn commented on public_html/wp-content/mu-plugins/0-error-handling.php:
```diff
@@ -0,0 +1,169 @@
+	$data = array(
+		'last_reported_at' => time(),
+		'error_count'      => 0, // since last reported.
+	);
```
low priority, but it'd be nice to remove those blank lines and align the `=` operator.

iandunn commented on public_html/wp-content/mu-plugins/0-error-handling.php:
```diff
@@ -0,0 +1,169 @@
+		}
+	}
+
+	$domain    = get_site_url();
```
If we move the file earlier, this (and similar functions) might not exist yet either.

iandunn commented on public_html/wp-content/mu-plugins/0-error-handling.php:
```diff
@@ -0,0 +1,169 @@
+
+/**
+ * Check and create filesystem dirs to manage rate limiting in error handling.
+ * For legacy bugs we are doing rate limiting via filesystem. We would be investigating to see if we can instead use memcache to rate limit sometime in the future.
```
Nitpicky, but there should be an empty line here.

iandunn commented on public_html/wp-content/mu-plugins/0-error-handling.php:
```diff
@@ -0,0 +1,169 @@
+ */
+function init_error_handling() {
+	if ( ! file_exists( ERROR_RATE_LIMITING_DIR ) ) {
+		mkdir( ERROR_RATE_LIMITING_DIR );
```
`wp_mkdir_p()` would be a best practice, but probably not helpful in this case, since `/tmp` should always exist and we wouldn't have permission to create it if it didn't.

iandunn commented on public_html/wp-content/mu-plugins/0-error-handling.php:
```diff
@@ -0,0 +1,169 @@
+}
+
+/**
+ * Function `send_error_to_slack` above also creates a bunch of files in /tmp/error_limiting folder in order to rate limit the notification.
```
This is missing a short doc.

iandunn reviewed (approved):
Mostly nitpicky comments, feel free to ignore :)

It'd be really nice to get the handler setup as early in the execution path as possible, though. Maybe this could be split into two files? One that loads super early and just logs the errors to disk, and a "phase 2" that runs at shutdown and sends to Slack, removes files, etc?

That's not essential if it's gonna be a ton of time, though. In that case maybe just leave a comment at the top that it's a future interation. We'll want to remind ourselves that it's possible for errors to occur early and not be logged, in case we're trying to debug something and can't figure out what's going on.

coreymckrill committed:
Add todo for splitting functionality, code cleanup

coreymckrill committed:
Change layout of messages sent to Slack, safelist more error types

This makes the error messages a bit easier to parse when scanning
quickly through Slack backscroll. It also differentiates the severity
of the error by attachment color.

This also adds more error types to the safelist. We can comment some
of these out if it becomes too noisy again or we decide we don't
care about some types, notices for example.

coreymckrill commented:
Committed in https://meta.trac.wordpress.org/changeset/8287
